File: src/main/java/chapter3/MapExample.java
package chapter3;

import io.reactivex.Observable;
import io.reactivex.functions.Function;


public class MapExample {

    public void marbleDiagram() {
        String[] balls = {"1", "2", "3", "5"};
        Observable<String> source = Observable.fromArray(balls)
                .map(ball -> ball + "<>");
        source.subscribe(System.out::println);
    }

    public void mapFunction() {
        Function<String, String> getDiamond = ball -> ball + "<>";

        String[] balls = {"1", "2", "3", "5"};
        Observable<String> source = Observable.fromArray(balls)
                .map(getDiamond);
        source.subscribe(System.out::println);
    }

    public void mappingType() {
        Function<String, Integer> ballToIndex = ball -> {
            switch (ball) {
                case "RED":
                    return 1;
                case "YELLOW":
                    return 2;
                case "GREEN":
                    return 3;
                case "BLUE":
                    return 5;
                default:
                    return -1;
            }
        };

        String[] balls = {"RED", "YELLOW", "GREEN", "BLUE"};
        Observable<Integer> source = Observable.fromArray(balls)
                .map(ballToIndex);
        source.subscribe(System.out::println);
    }

    public static void main(String[] args) {
        MapExample demo = new MapExample();
        demo.marbleDiagram();
        demo.mapFunction();
        demo.mappingType();
    }

}

File: src/main/kotlin/chapter3/MapExample_.kt
package chapter3

import io.reactivex.Observable


class MapExample_ {

    fun marbleDiagram() {
        val balls = arrayOf("1", "2", "3", "5")
        val source = Observable.fromArray(*balls)
                .map { it + "<>" }
        source.subscribe { println(it) }
    }

    fun mapFunction() {
        val getDiamond = { ball: String -> ball + "<>" }

        val balls = arrayOf("1", "2", "3", "5")
        /*
        val source = Observable.fromArray(*balls)
                .map(getDiamond)
        */
        val source = Observable.fromArray(*balls)
                .map { getDiamond(it) }
        source.subscribe { println(it) }
    }

    fun mappingType() {
        val ballToIndex = { ball: String ->
            when (ball) {
                "RED" -> 1
                "YELLOW" -> 2
                "GREEN" -> 3
                "BLUE" -> 5
                else -> -1
            }
        }

        val balls = arrayOf("RED", "YELLOW", "GREEN", "BLUE")
        val source = Observable.fromArray(*balls)
                .map { ballToIndex(it) }
        source.subscribe { println(it) }
    }

}


fun main(args: Array<String>) {
    val demo = MapExample_()
    demo.marbleDiagram()
    demo.mapFunction()
    demo.mappingType()
}